destination/writer.go

```diff
@@ -0,0 +1,226 @@
+// Copyright © 2025 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package destination
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"cloud.google.com/go/spanner"
+	"github.com/conduitio/conduit-commons/opencdc"
+)
+
+type ColumnTypes map[string]string
+
+// Writer implements a writer logic for Spanner destination.
+type Writer struct {
+	client *spanner.Client
+	// Function to dynamically get table name for each record.
+	tableNameFunc TableFn
+	// Maps table names to their column types.
+	columnTypes map[string]ColumnTypes
+	// Schema name to write data into.
+	schema string
+}
+
+// NewWriter creates new instance of the Writer.
+func NewWriter(_ context.Context, client *spanner.Client, config Config) (*Writer, error) {
+	tableFn, err := config.TableFunction()
+	if err != nil {
+		return nil, fmt.Errorf("invalid table name or table function: %w", err)
+	}
+
+	writer := &Writer{
+		client:        client,
+		tableNameFunc: tableFn,
+		schema:        config.Schema,
+		columnTypes:   make(map[string]ColumnTypes),
+	}
+
+	return writer, nil
+}
+
+// Insert inserts a record.
+func (w *Writer) Insert(ctx context.Context, record opencdc.Record) error {
+	table, err := w.tableNameFunc(record)
+	if err != nil {
+		return err
+	}
+
+	payload, err := w.structurizeData(record.Payload.After)
+	if err != nil {
+		return fmt.Errorf("structurize payload: %w", err)
+	}
+	if len(payload) == 0 {
```

Comment:
is the key not enough to insert a record? because even if the payload is empty, we still have rec.Key

Reply:
Yes, updated.